datafusion/datasource/src/source.rs

```diff
@@ -79,6 +79,13 @@ pub trait DataSource: Send + Sync + Debug {
         &self,
         _projection: &ProjectionExec,
     ) -> datafusion_common::Result<Option<Arc<dyn ExecutionPlan>>>;
+
+    fn push_down_filter(
```

Comment:
Can you please document what the expectations for this function are? Specifically, I think it would be great to answer the following questions:

1. If this returns `Some(..)` does that *guarantee* that any rows that don't pass the filter are filtered out? Or is it just a best effort?
2. Should the plan recursively push the filter into its input (I think the answer is yes)